File: src/tokens.ts
import type { TokenMethod } from './types.ts';

// 'tiktoken' — uses gpt-tokenizer (cl100k_base / tiktoken), exact for GPT-4 and GPT-3.5-turbo.
// Also a reliable proxy for Anthropic Claude models: token density is comparable for
// English prose (typically within 5-10%). No official Claude tokenizer is available
// outside the Anthropic API, so this is the best local option for Claude token counts.
let gptEncode: ((text: string) => number[]) | null = null;
try {
  const mod = await import('gpt-tokenizer');
  gptEncode = mod.encode;
} catch { /* not installed; 'tiktoken' tokenMethod will throw if used */ }

export function estimateTokens(text: string, method: TokenMethod = 'chars'): number {
  if (method === 'tiktoken') {
    if (!gptEncode) throw new Error('gpt-tokenizer not installed. Run: bun add gpt-tokenizer');
    return gptEncode(text).length;
  }
  // 'chars': fast zero-dependency approximation (~±20% vs real token count)
  return Math.ceil(text.length / 4);
}

File: crates/terse/src/tokens.rs
use crate::types::TokenMethod;
use once_cell::sync::Lazy;
use tiktoken_rs::CoreBPE;

// cl100k_base: GPT-4 / text-embedding-ada-002.
// Good proxy for Claude: token density is comparable for English prose (typically ±5-10%).
static CL100K: Lazy<CoreBPE> =
    Lazy::new(|| tiktoken_rs::cl100k_base().expect("failed to load cl100k_base tokenizer"));

pub fn estimate_tokens(text: &str, method: &TokenMethod) -> Result<usize, String> {
    match method {
        TokenMethod::Chars => Ok((text.len() + 3) / 4),
        TokenMethod::Tiktoken => Ok(CL100K.encode_with_special_tokens(text).len()),
    }
}
